temporalio/worker/replayer.py

```diff
@@ -0,0 +1,272 @@
+
+        # Extract workflow started event
+        started_event = next(
+            (
+                e
+                for e in history.events
+                if e.HasField("workflow_execution_started_event_attributes")
+            ),
+            None,
+        )
+        if not started_event:
+            raise ValueError("Started event not found")
+
+        # Create bridge worker and workflow worker
+        bridge_worker = temporalio.bridge.worker.Worker.for_replay(
+            history,
+            temporalio.bridge.worker.WorkerConfig(
+                namespace=self._config["namespace"],
+                task_queue=started_event.workflow_execution_started_event_attributes.task_queue.name,
+                build_id=self._config["build_id"] or load_default_build_id(),
+                identity_override=self._config["identity"],
+                # All values below are ignored but required by Core
+                max_cached_workflows=1,
+                max_outstanding_workflow_tasks=1,
+                max_outstanding_activities=1,
+                max_outstanding_local_activities=1,
+                max_concurrent_wft_polls=1,
+                nonsticky_to_sticky_poll_ratio=1,
+                max_concurrent_at_polls=1,
+                no_remote_activities=True,
+                sticky_queue_schedule_to_start_timeout_millis=1000,
+                max_heartbeat_throttle_interval_millis=1000,
+                default_heartbeat_throttle_interval_millis=1000,
+            ),
+        )
+        workflow_worker = _WorkflowWorker(
+            bridge_worker=lambda: bridge_worker,
+            namespace=self._config["namespace"],
+            task_queue=started_event.workflow_execution_started_event_attributes.task_queue.name,
+            workflows=self._config["workflows"],
+            workflow_task_executor=self._config["workflow_task_executor"],
+            workflow_runner=self._config["workflow_runner"],
+            data_converter=self._config["data_converter"],
+            interceptors=self._config["interceptors"],
+            debug_mode=self._config["debug_mode"],
+            fail_on_eviction=True,
+        )
+
+        # Run it
+        try:
+            await workflow_worker.run()
+        finally:
+            # We must finalize shutdown here
+            try:
+                await bridge_worker.finalize_shutdown()
+            except Exception:
+                logger.warning("Failed to finalize shutdown", exc_info=True)
+
+
+class ReplayerConfig(TypedDict, total=False):
+    """TypedDict of config originally passed to :py:class:`Replayer`."""
+
+    workflows: Iterable[Type]
+    workflow_task_executor: Optional[concurrent.futures.ThreadPoolExecutor]
+    workflow_runner: WorkflowRunner
+    namespace: str
+    data_converter: temporalio.converter.DataConverter
+    interceptors: Iterable[Interceptor]
+    build_id: Optional[str]
+    identity: Optional[str]
+    debug_mode: bool
+
+
+def _history_from_json(
```

Comment:
Wow I hate this so much.

We should really, really just change tctl & web to stop emitting this nonsense. I keep forgetting why we haven't / won't, probably because it's a lame reason. We should bite the bullet and do that. Putting this in every SDK is beyond annoying.

If we really can't for some dumb reason, let's just put it in core and I'll expose a static function to do the transform.

Reply:
> probably because it's a lame reason.

Yup. There are compatibility concerns because the SDK's blindly convert in a non-idempotent way. That reminds me, I need to make this idempotent so if it was already fixed it'd be ok. But if we fixed this in the server, it'd break our uses of it.

> let's just put it in core and I'll expose a static function to do the transform

That'd be fine. We are not exposing this helper to users, so a core function that took raw JSON and made a Rust history proto directly that we could pass to `init_replay_worker` would work (or an `init_replay_worker_json` or something).